assign from binary method panic

The following program `sample.go` triggers a panic:

```go
package main

import (
    "io/ioutil"
    "os"
)

func main() {
    file, err := ioutil.TempFile("", "yeagibench")
    if err != nil {
        panic(err)
    }

    _, err = file.Write([]byte("hello world"))
    if err != nil {
        panic(err)
    }

    err = file.Close()
    if err != nil {
        panic(err)
    }

    _, err = ioutil.ReadFile(file.Name())
    if err != nil {
        panic(err)
    }

    err = os.Remove(file.Name())
    if err != nil {
        panic(err)
    }
}
```

Expected result:
```console
$ go run ./sample.go
```

Got:
```console
$ yaegi ./sample.go
panic: reflect.Set: value of type int is not assignable to type func(string, string) (*os.File, error) [recovered]
	panic: reflect.Set: value of type int is not assignable to type func(string, string) (*os.File, error)

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc0000b7720)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:107 +0x134
panic(0x17a87e0, 0xc00016d210)
	/usr/local/Cellar/go/1.12.3/libexec/src/runtime/panic.go:522 +0x1b5
reflect.Value.assignTo(0x17a76a0, 0xc00015a120, 0x82, 0x18b74f6, 0xb, 0x17e99a0, 0x0, 0x50, 0x0, 0x0)
	/usr/local/Cellar/go/1.12.3/libexec/src/reflect/value.go:2339 +0x437
reflect.Value.Set(0x17e99a0, 0xc0000ae828, 0x193, 0x17a76a0, 0xc00015a120, 0x82)
	/usr/local/Cellar/go/1.12.3/libexec/src/reflect/value.go:1473 +0xa8
github.com/containous/yaegi/interp.assignX.func1(0xc0000b7720, 0xc0001c68c0)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:206 +0x2c4
github.com/containous/yaegi/interp.runCfg(0xc0000f70e0, 0xc0000b7720)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:112 +0x62
github.com/containous/yaegi/interp.(*Interpreter).run(0xc0000b05a0, 0xc0000f6780, 0xc0000b7590)
	/Users/marc/go/src/github.com/containous/yaegi/interp/run.go:94 +0x126
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc0000b05a0, 0xc0001c2000, 0x19b, 0x39b, 0xc0001c2000, 0x19b, 0x0, 0xc0001a6820)
	/Users/marc/go/src/github.com/containous/yaegi/interp/interp.go:269 +0x43c
main.main()
	/Users/marc/go/src/github.com/containous/yaegi/yaegi.go:59 +0x3ec
```

--- a/_test/file_access.go
+++ b/_test/file_access.go
@@ -1,0 +1,40 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+)
+
+func main() {
+	file, err := ioutil.TempFile("", "yeagibench")
+	if err != nil {
+		panic(err)
+	}
+
+	n, err := file.Write([]byte("hello world"))
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println("n:", n)
+
+	err = file.Close()
+	if err != nil {
+		panic(err)
+	}
+
+	b, err := ioutil.ReadFile(file.Name())
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println("b:", string(b))
+
+	err = os.Remove(file.Name())
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Output:
+// n: 11
+// b: hello world

--- a/_test/ret6.go
+++ b/_test/ret6.go
@@ -1,0 +1,15 @@
+package main
+
+import "fmt"
+
+type Foo struct{}
+
+func foo() *Foo { return nil }
+
+func main() {
+	f := foo()
+	fmt.Println(f)
+}
+
+// Output:
+// <nil>

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -309,7 +309,7 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 					// Do not overload existings symbols (defined in GTA) in global scope
 					sym, _, _ = scope.lookup(dest.ident)
 				} else {
-					sym = &Symbol{index: scope.add(dest.typ), kind: Var, global: scope.global}
+					sym = &Symbol{index: scope.add(dest.typ), kind: Var}
 					scope.sym[dest.ident] = sym
 				}
 				dest.val = src.val
@@ -352,6 +352,10 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 			// Propagate type
 			// TODO: Check that existing destination type matches source type
 			switch {
+			case src.action == Call:
+				n.gen = nop
+				src.level = level
+				src.findex = dest.findex
 			case src.action == Recv:
 				// Assign by reading from a receiving channel
 				n.gen = nop
@@ -402,6 +406,8 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 			wireChild(n)
 			l := len(n.child) - 1
 			switch n.child[l].kind {
+			case CallExpr:
+				n.gen = nop
 			case IndexExpr:
 				n.child[l].gen = getIndexMap2
 				n.gen = nop
@@ -429,9 +435,7 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 				} else {
 					types = funtype.ret
 				}
-				if l > len(types) {
-					n.gen = assignX2
-				}
+				n.gen = nop
 
 			case IndexExpr:
 				types = append(types, n.child[l].child[0].typ.val, scope.getType("bool"))
@@ -785,7 +789,7 @@ func (interp *Interpreter) Cfg(root *Node) ([]*Node, error) {
 			wireChild(n)
 
 		case Ident:
-			if isKey(n) || isNewDefine(n) {
+			if isKey(n) || isNewDefine(n, scope) {
 				break
 			} else if sym, level, ok := scope.lookup(n.ident); ok {
 				// Found symbol, populate node info
@@ -1330,7 +1334,8 @@ func isKey(n *Node) bool {
 		(n.anc.kind == KeyValueExpr && n.anc.child[0] == n)
 }
 
-func isNewDefine(n *Node) bool {
+// isNewDefine returns true if node refers to a new definition
+func isNewDefine(n *Node, scope *Scope) bool {
 	if n.ident == "_" {
 		return true
 	}
@@ -1340,8 +1345,14 @@ func isNewDefine(n *Node) bool {
 	if n.anc.kind == DefineX && n.anc.lastChild() != n {
 		return true
 	}
-	if n.anc.kind == RangeStmt && (n.anc.child[0] == n || n.anc.child[1] == n) {
-		return true
+	if n.anc.kind == RangeStmt {
+		if n.anc.child[0] == n {
+			return true // array or map key, or chan element
+		}
+		if scope.rangeChanType(n.anc) == nil && n.anc.child[1] == n {
+			return true // array or map value
+		}
+		return false // array, map or channel are always pre-defined in range expression
 	}
 	if n.anc.kind == ValueSpec && n.anc.lastChild() != n {
 		return true

--- a/interp/interp_eval_test.go
+++ b/interp/interp_eval_test.go
@@ -142,7 +142,7 @@ func TestEvalNil(t *testing.T) {
 				`)
 			},
 			src: "Hello()",
-			res: "<invalid reflect.Value>",
+			res: "<nil>",
 		},
 	})
 }

--- a/interp/run.go
+++ b/interp/run.go
@@ -17,7 +17,6 @@ var builtin = [...]BuiltinGenerator{
 	Nop:          nop,
 	Addr:         addr,
 	Assign:       assign,
-	AssignX:      assignX,
 	Add:          add,
 	AddAssign:    addAssign,
 	And:          and,
@@ -187,48 +186,6 @@ func convert(n *Node) {
 	}
 }
 
-// assignX implements multiple value assignment
-func assignX(n *Node) {
-	l := len(n.child) - 1
-	b := n.child[l].findex
-	s := n.child[:l]
-	next := getExec(n.tnext)
-	values := make([]func(*Frame) reflect.Value, l)
-	for i, c := range s {
-		values[i] = genValue(c)
-	}
-
-	n.exec = func(f *Frame) Builtin {
-		for i, value := range values {
-			if f.data[b+i].IsValid() {
-				value(f).Set(f.data[b+i])
-			}
-		}
-		return next
-	}
-}
-
-// assignX2 implements multiple value assignment for expression where type is defined
-func assignX2(n *Node) {
-	l := len(n.child) - 2
-	b := n.child[l].findex
-	s := n.child[:l]
-	next := getExec(n.tnext)
-	values := make([]func(*Frame) reflect.Value, l)
-	for i, c := range s {
-		values[i] = genValue(c)
-	}
-
-	n.exec = func(f *Frame) Builtin {
-		for i, value := range values {
-			if f.data[b+i].IsValid() {
-				value(f).Set(f.data[b+i])
-			}
-		}
-		return next
-	}
-}
-
 // assign implements single value assignment
 func assign(n *Node) {
 	next := getExec(n.tnext)
@@ -368,15 +325,15 @@ func _println(n *Node) {
 
 func _recover(n *Node) {
 	tnext := getExec(n.tnext)
-	i := n.findex
+	dest := genValue(n)
 	var err error
-	nilErr := reflect.ValueOf(&err).Elem()
+	nilErr := reflect.ValueOf(valueInterface{n, reflect.ValueOf(&err).Elem()})
 
 	n.exec = func(f *Frame) Builtin {
 		if f.anc.recovered == nil {
-			f.data[i] = nilErr
+			dest(f).Set(nilErr)
 		} else {
-			f.data[i] = reflect.ValueOf(f.anc.recovered)
+			dest(f).Set(reflect.ValueOf(valueInterface{n, reflect.ValueOf(f.anc.recovered)}))
 			f.anc.recovered = nil
 		}
 		return tnext
@@ -541,6 +498,21 @@ func call(n *Node) {
 	for i := range n.child[0].typ.ret {
 		ret[i] = n.findex + i
 	}
+	rvalues := make([]func(*Frame) reflect.Value, len(n.child[0].typ.ret))
+	switch n.anc.kind {
+	case Define, AssignStmt, DefineX, AssignXStmt:
+		for i := range rvalues {
+			c := n.anc.child[i]
+			if c.ident != "_" {
+				rvalues[i] = genValue(c)
+			}
+		}
+	default:
+		for i := range n.child[0].typ.ret {
+			j := n.findex + i
+			rvalues[i] = func(f *Frame) reflect.Value { return f.data[j] }
+		}
+	}
 
 	n.exec = func(f *Frame) Builtin {
 		def := value(f).Interface().(*Node)
@@ -614,8 +586,10 @@ func call(n *Node) {
 			return fnext
 		}
 		// Propagate return values to caller frame
-		for i, r := range ret {
-			f.data[r] = nf.data[i]
+		for i, v := range rvalues {
+			if v != nil {
+				v(f).Set(nf.data[i])
+			}
 		}
 		return tnext
 	}
@@ -704,13 +678,38 @@ func callBin(n *Node) {
 			return fnext
 		}
 	default:
-		n.exec = func(f *Frame) Builtin {
-			in := make([]reflect.Value, l)
-			for i, v := range values {
-				in[i] = v(f)
+		switch n.anc.kind {
+		case Define, AssignStmt, DefineX, AssignXStmt:
+			rvalues := make([]func(*Frame) reflect.Value, funcType.NumOut())
+			for i := range rvalues {
+				c := n.anc.child[i]
+				if c.ident != "_" {
+					rvalues[i] = genValue(c)
+				}
 			}
-			copy(f.data[n.findex:], value(f).Call(in))
-			return tnext
+			n.exec = func(f *Frame) Builtin {
+				in := make([]reflect.Value, l)
+				for i, v := range values {
+					in[i] = v(f)
+				}
+				out := value(f).Call(in)
+				for i, v := range rvalues {
+					if v != nil {
+						v(f).Set(out[i])
+					}
+				}
+				return tnext
+			}
+		default:
+			n.exec = func(f *Frame) Builtin {
+				in := make([]reflect.Value, l)
+				for i, v := range values {
+					in[i] = v(f)
+				}
+				out := value(f).Call(in)
+				copy(f.data[n.findex:], out)
+				return tnext
+			}
 		}
 	}
 }
@@ -1328,7 +1327,7 @@ func appendSlice(n *Node) {
 }
 
 func _append(n *Node) {
-	i := n.findex
+	dest := genValue(n)
 	value := genValue(n.child[1])
 	next := getExec(n.tnext)
 
@@ -1345,38 +1344,38 @@ func _append(n *Node) {
 			for i, v := range values {
 				sl[i] = v(f)
 			}
-			f.data[i] = reflect.Append(value(f), sl...)
+			dest(f).Set(reflect.Append(value(f), sl...))
 			return next
 		}
 	} else {
 		value0 := genValue(n.child[2])
 
 		n.exec = func(f *Frame) Builtin {
-			f.data[i] = reflect.Append(value(f), value0(f))
+			dest(f).Set(reflect.Append(value(f), value0(f)))
 			return next
 		}
 	}
 }
 
 func _cap(n *Node) {
-	i := n.findex
+	dest := genValue(n)
 	value := genValue(n.child[1])
 	next := getExec(n.tnext)
 
 	n.exec = func(f *Frame) Builtin {
-		f.data[i].SetInt(int64(value(f).Cap()))
+		dest(f).SetInt(int64(value(f).Cap()))
 		return next
 	}
 }
 
 func _copy(n *Node) {
-	i := n.findex
+	dest := genValue(n)
 	value0 := genValue(n.child[1])
 	value1 := genValue(n.child[2])
 	next := getExec(n.tnext)
 
 	n.exec = func(f *Frame) Builtin {
-		f.data[i].SetInt(int64(reflect.Copy(value0(f), value1(f))))
+		dest(f).SetInt(int64(reflect.Copy(value0(f), value1(f))))
 		return next
 	}
 }
@@ -1465,7 +1464,7 @@ func _new(n *Node) {
 
 // _make allocates and initializes a slice, a map or a chan.
 func _make(n *Node) {
-	i := n.findex
+	dest := genValue(n)
 	next := getExec(n.tnext)
 	typ := n.child[1].typ.TypeOf()
 
@@ -1477,13 +1476,13 @@ func _make(n *Node) {
 		case 3:
 			n.exec = func(f *Frame) Builtin {
 				len := int(value(f).Int())
-				f.data[i] = reflect.MakeSlice(typ, len, len)
+				dest(f).Set(reflect.MakeSlice(typ, len, len))
 				return next
 			}
 		case 4:
 			value1 := genValue(n.child[3])
 			n.exec = func(f *Frame) Builtin {
-				f.data[i] = reflect.MakeSlice(typ, int(value(f).Int()), int(value1(f).Int()))
+				dest(f).Set(reflect.MakeSlice(typ, int(value(f).Int()), int(value1(f).Int())))
 				return next
 			}
 		}
@@ -1492,13 +1491,13 @@ func _make(n *Node) {
 		switch len(n.child) {
 		case 2:
 			n.exec = func(f *Frame) Builtin {
-				f.data[i] = reflect.MakeChan(typ, 0)
+				dest(f).Set(reflect.MakeChan(typ, 0))
 				return next
 			}
 		case 3:
 			value := genValue(n.child[2])
 			n.exec = func(f *Frame) Builtin {
-				f.data[i] = reflect.MakeChan(typ, int(value(f).Int()))
+				dest(f).Set(reflect.MakeChan(typ, int(value(f).Int())))
 				return next
 			}
 		}
@@ -1507,13 +1506,13 @@ func _make(n *Node) {
 		switch len(n.child) {
 		case 2:
 			n.exec = func(f *Frame) Builtin {
-				f.data[i] = reflect.MakeMap(typ)
+				dest(f).Set(reflect.MakeMap(typ))
 				return next
 			}
 		case 3:
 			value := genValue(n.child[2])
 			n.exec = func(f *Frame) Builtin {
-				f.data[i] = reflect.MakeMapWithSize(typ, int(value(f).Int()))
+				dest(f).Set(reflect.MakeMapWithSize(typ, int(value(f).Int())))
 				return next
 			}
 		}

--- a/interp/type.go
+++ b/interp/type.go
@@ -423,7 +423,7 @@ func (t *Type) zero() (v reflect.Value, err error) {
 	case AliasT:
 		v, err = t.val.zero()
 
-	case ArrayT, StructT:
+	case ArrayT, PtrT, StructT:
 		v = reflect.New(t.TypeOf()).Elem()
 
 	case ValueT:

